BLE Indoor Positioning/src/test/java/com/nexenio/bleindoorpositioning/ble/advertising/AdvertisingPacketFactoryTest.java

```diff
@@ -0,0 +1,161 @@
+        assertTrue(advertisingPacketFactory.canCreateAdvertisingPacketWithSubFactories(testData));
+    }
+
+    @Test
+    public void addAdvertisingPacketFactory_incorrectType_addsNothing() {
+        IBeaconAdvertisingPacketFactory advertisingPacketFactory = new IBeaconAdvertisingPacketFactory();
+        AdvertisingPacketFactory testFactory = new TestAdvertisingPacketFactory();
+        advertisingPacketFactory.addAdvertisingPacketFactory(testFactory);
+
+        assertEquals(0, advertisingPacketFactory.getSubFactoryMap().size());
+        assertEquals(null, advertisingPacketFactory.getAdvertisingPacketFactory(TestAdvertisingPacket.class));
+    }
+
+    @Test
+    public void removeAdvertisingPacketFactory_existingFactory_removesCorrectly() {
+        EddystoneAdvertisingPacketFactory advertisingPacketFactory = new EddystoneAdvertisingPacketFactory();
+        AdvertisingPacketFactory testFactory = new TestAdvertisingPacketFactory();
+        advertisingPacketFactory.addAdvertisingPacketFactory(testFactory);
+
+        advertisingPacketFactory.removeAdvertisingPacketFactory(testFactory);
+        assertEquals(0, advertisingPacketFactory.getSubFactoryMap().size());
+        assertEquals(null, advertisingPacketFactory.getAdvertisingPacketFactory(TestAdvertisingPacket.class));
+    }
+
+    @Test
+    public void removeAdvertisingPacketFactory_noExistingFactory_removesCorrectly() {
+        EddystoneAdvertisingPacketFactory advertisingPacketFactory = new EddystoneAdvertisingPacketFactory();
+        AdvertisingPacketFactory testFactory = new TestAdvertisingPacketFactory();
+
+        advertisingPacketFactory.removeAdvertisingPacketFactory(testFactory);
+        assertEquals(0, advertisingPacketFactory.getSubFactoryMap().size());
+        assertEquals(null, advertisingPacketFactory.getAdvertisingPacketFactory(TestAdvertisingPacket.class));
+    }
+
+    @Test
+    public void canCreateAdvertisingPacket_validData_returnsTrue() {
+        IBeaconAdvertisingPacketFactory advertisingPacketFactory = new IBeaconAdvertisingPacketFactory();
+        assertTrue(advertisingPacketFactory.canCreateAdvertisingPacketWithSubFactories(BeaconTest.IBEACON_ADVERTISING_DATA));
+    }
+
+    @Test
+    public void canCreateAdvertisingPacket_invalidData_returnsFalse() {
+        IBeaconAdvertisingPacketFactory advertisingPacketFactory = new IBeaconAdvertisingPacketFactory();
+        assertFalse(advertisingPacketFactory.canCreateAdvertisingPacket(new byte[5]));
+    }
+
+    @Test
+    public void canCreateAdvertisingPacketWithSubFactories_validData_returnsTrue() {
+        EddystoneAdvertisingPacketFactory advertisingPacketFactory = new EddystoneAdvertisingPacketFactory();
+        AdvertisingPacketFactory testFactory = new TestAdvertisingPacketFactory();
+        advertisingPacketFactory.addAdvertisingPacketFactory(testFactory);
+
+        byte[] testData = new byte[5];
+        assertFalse(advertisingPacketFactory.canCreateAdvertisingPacket(testData));
```

Comment:
May be this assertion is not needed here. It seems a little confusing because validData is in the signature, but why is new byte[5] valid here and invalid above?

Reply:
Because the `EddystoneAdvertisingPacketFactory` should not be able to create an advertising packet for that data. The `TestAdvertisingPacketFactory` however should be able to do so, which is a sub-factory of `advertisingPacketFactory`.